---
title: "scRNA-Seq Cell_Type_Annotation"
author: "Tapan"
date: "2023-06-28"
output: html_notebook
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Cell Type Annotation

This is a basic tutorial for cell type annotation using singleR
## Install R packages
```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("SingleR")
```

```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("biomaRt")
```


```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("SingleCellExperiment")

install.packages('Seurat')

install.packages("dplyr")
install.packages('tidyverse')
install.packages('scater')
```
## Function to convert ensemble ID to Gene Symbol
```{r}
library(tidyverse)
convert_to_Symbol <- function(obj){
  library(biomaRt)
  if( any(str_detect(rownames(obj), "ENSG")) & any(str_detect(rownames(obj), "ENSMU")))
  {
   
    ensemble_h <- useMart("ensembl", dataset = c( 'hsapiens_gene_ensembl'))
    gene_info_h <- getBM(attributes=c( "ensembl_gene_id", 'hgnc_symbol'), filters="ensembl_gene_id", values= rownames(obj), mart=ensemble_h)
    ensemble_m <- useMart("ensembl", dataset = c( 'mmusculus_gene_ensembl'))
    gene_info_m <- getBM(attributes=c( "ensembl_gene_id", 'external_gene_name'), filters="ensembl_gene_id", values= rownames(obj), mart=ensemble_m)
    colnames(gene_info_m)[2] <- 'hgnc_symbol'
    names<- as.data.frame(rownames(obj))
    colnames(names)[1] <- 'ensembl_gene_id'
    symbol=rbind(gene_info_m, gene_info_h)
    dt<- merge(names, symbol, all= TRUE)
    merge_dt <- aggregate(dt$hgnc_symbol, by = list(dt$ensembl_gene_id), FUN = paste0, collapse = ",") # merge multiple gene Symbol for same ID by comma
    colnames(merge_dt) <- colnames(dt)
  }
  else if(any(str_detect(rownames(obj), "ENSG"))==TRUE & any(str_detect(rownames(obj), "ENSMU")) == FALSE) {
    ensemble_h <- useMart("ensembl", dataset = c( 'hsapiens_gene_ensembl'))
    gene_info_h <- getBM(attributes=c( "ensembl_gene_id", 'hgnc_symbol'), filters="ensembl_gene_id", values= rownames(obj), mart=ensemble_h)
    names<- as.data.frame(rownames(obj))
    colnames(names)[1] <- 'ensembl_gene_id'
    dt<- merge(names, gene_info_h, all= TRUE)
    merge_dt <- aggregate(dt$hgnc_symbol, by = list(dt$ensembl_gene_id), FUN = paste0, collapse = ",")
    colnames(merge_dt) <- colnames(dt)
    
  }
  else if(any(str_detect(rownames(obj), "ENSG"))==FALSE & any(str_detect(rownames(obj), "ENSMU")) == TRUE) {
    ensemble_m <- useMart("ensembl", dataset = c( 'mmusculus_gene_ensembl'))
    gene_info_m <- getBM(attributes=c( "ensembl_gene_id", 'external_gene_name'), filters="ensembl_gene_id", values= rownames(obj), mart=ensemble_m)
    colnames(gene_info_m)[2] <- 'hgnc_symbol'
    names<- as.data.frame(rownames(obj))
    colnames(names)[1] <- 'ensembl_gene_id'
    dt<- merge(names, gene_info_m, all= TRUE)
    merge_dt <- aggregate(dt$hgnc_symbol, by = list(dt$ensembl_gene_id), FUN = paste0, collapse = ",")
    colnames(merge_dt) <- colnames(dt)
    
  }
  rowData(obj)$ENSEMBL <- rownames(obj)
  rowData(obj)$SYMBOL <- merge_dt$hgnc_symbol
  new.names <- rowData(obj)$SYMBOL
  new.names[new.names == 'NA'] <- NA
  missing.name <- is.na(new.names)
  new.names[missing.name] <- rowData(obj)$ENSEMBL[missing.name]
  dup.name <- new.names %in% new.names[duplicated(new.names)]
  new.names[dup.name] <- paste0(new.names, "_", rowData(obj)$ENSEMBL)[dup.name]
  rownames(obj) <- new.names
  
  return(obj)
}
```

```{r}
## 10X data input and convert to SingleCellExperiment object
#Load 10X Data 
raw_data<- function(Data.path){
  Raw_data <- Seurat::Read10X(data.dir = Data.path)
  sce<- SingleCellExperiment::SingleCellExperiment(assay = list(counts = data))
  return(sce)
}
#convert Ensemble ID to gene Symbol
 add_gene_symbol<- function(sce){
  if( any(str_detect(rownames(sce), "ENSG")) | any(str_detect(rownames(sce), "ENSMU"))){
      sce <- convert_to_Symbol(sce)
    }
    else{
      sce<- sce
    }
  return(sce)
 }
 
# Function to add QC Statistics to the sce object
add_QC_stats <- function(sce){
  
#calculate %MT genes
  mito_genes <- rownames(sce)[grep("(?i)^MT-", rownames(sce))]
  # Ribosomal genes
  ribo_genes <- rownames(sce)[grep("(?i)^RP[SL]", rownames(sce))]
  
  # Hemoglobin genes - includes all genes starting with HB except HBP.
  hb_genes <- rownames(sce)[grep("(?i)^HB[^(P)]", rownames(sce))]
  
  #add QC
  sce <- scuttle::addPerCellQC(sce, flatten = T, subsets = list(mt = mito_genes, hb = hb_genes, ribo = ribo_genes))
  
  #rename colnames
  if( any(colnames(colData(sce))=="Cell")){
    sce@colData <- subset(sce@colData , select= -c(Cell))
  }
 
  colTable=colData(sce) %>% as.data.frame()  %>% 
  dplyr::select(-c("total",  "subsets_mt_sum","subsets_mt_detected","subsets_hb_sum","subsets_hb_detected", "subsets_ribo_sum","subsets_ribo_detected" ))  %>% 
  dplyr::rename(nCount_RNA=sum, nFeature_RNA=detected, Mito_gene_percent=subsets_mt_percent, Hemoglobin_gene_percent=subsets_hb_percent, Ribosomal_gene_percent= subsets_ribo_percent)
  
  sce<- SingleCellExperiment(assay = list(counts = assay(sce,'counts')),
                             colData = colTable)                                               
  sce@colData$Hemoglobin_gene_percent <- replace(sce@colData$Hemoglobin_gene_percent, is.na(sce@colData$Hemoglobin_gene_percent), 0)
  sce@colData$Mito_gene_percent <- replace(sce@colData$Mito_gene_percent, is.na(sce@colData$Mito_gene_percent), 0)
  sce@colData$Ribosomal_gene_percent <- replace(sce@colData$Ribosomal_gene_percent, is.na(sce@colData$Ribosomal_gene_percent), 0)
return(sce)
  
}
```

```{r}
# Raw data filter (QC)
filter_raw_data <- function(
    sce, #SingleCellExperiment object
    min_count_gene=3, #min times a gene is expressed
    max_count_gene=10000000, # max times a gene is expressed
    min_genes_per_cell=NULL, # min expressed genes per cell
    max_genes_per_cell=NULL, # max expressed genes per cell
    min_count_cell=NULL,  # min counts per cell
    max_count_cell=NULL,  # max counts per cell
    mt_percnt=0, #Mitochondrial gene percentage
    ribsoml_percnt=0 #Ribosomal gene percentage
    #hemglbn_percnt=NULL #Hemoglobin gene percentage
    
)

{
  # if(any(is.na(sce@colData$Patients)))
  # {
  #   sce@colData <- subset(sce@colData, select = -c(Patients,Type,Batch, batch))
  # }
  selected_f <- rownames(sce)[Matrix::rowSums(counts(sce)) >= min_count_gene & Matrix::rowSums(counts(sce)) < max_count_gene ]
  sce<-sce[selected_f]
  selected_ribo <- sce$Ribosomal_gene_percent >= ribsoml_percnt
  sce <- sce[,  selected_ribo]
  metadata<-sce %>%
    colData() %>% as.data.frame() %>% filter(nFeature_RNA>= min_genes_per_cell & nFeature_RNA< max_genes_per_cell & nCount_RNA>= min_count_cell 
                                             & nCount_RNA< max_count_cell & Ribosomal_gene_percent>= ribsoml_percnt & Mito_gene_percent <= mt_percnt ) 
  count<- assay(sce, "counts")
  colnames(count)<- sce@colData$Barcode
  count<-count[,colnames(count) %in% metadata$Barcode]
  
  sce<- SingleCellExperiment(assay = list(counts = count),
                             colData = data.frame(metadata))
  sce@colData$All<-"sample"
  return(sce)
  
}
```

## Function variable feature selection
```{r}
VariableFeatures = function(sce, used = 'counts', ngene = 1000, batch = F) {
  # use assay of interest to find top HVG
 
    edata = assay(sce, used)
    edata = log2(edata+1)

  if(batch) {
    batch = sce@colData$batch
    
    var = as.data.frame(scran::modelGeneVar(edata, block = batch))
    var=var %>% drop_na() 
    top.hvgs2 <- scran::getTopHVGs(var, n=ngene)
    var$Status<-is.element(rownames(var),top.hvgs2)
    top20<-var %>% filter(Status=="TRUE") %>% arrange(FDR) %>% arrange(desc(bio)) %>% head(20)
    selected_genes<-var %>% filter(Status=="TRUE") %>% arrange(FDR)
    colnames(selected_genes)[c(1,4)]<-c("Average_expression","Variance")
    edata.od = edata[top.hvgs2,]
    
    
  } else {
    var = as.data.frame(scran::modelGeneVar(edata))
    var=var %>% drop_na() 
    top.hvgs2 <- scran::getTopHVGs(var, n=ngene)
    var$Status<-is.element(rownames(var),top.hvgs2)
    top20<-var %>% filter(Status=="TRUE") %>% arrange(FDR) %>% arrange(desc(bio)) %>% head(20)
    selected_genes<-var %>% filter(Status=="TRUE") %>% arrange(FDR)
    colnames(selected_genes)[c(1,4)]<-c("Average_expression","Variance")
    edata.od = edata[top.hvgs2,]

  }
  altExp(sce, 'VGcounts') = SingleCellExperiment(assay = list(VGcounts = edata.od))
  if(any(c("nCount_RNA","nFeature_RNA", "Mito_gene_percent", "Hemoglobin_gene_percent", "Ribosomal_gene_percent") == colnames(colData(sce)))){
    sce@colData = subset(sce@colData, select = -c(nCount_RNA,nFeature_RNA, Mito_gene_percent, Hemoglobin_gene_percent, Ribosomal_gene_percent)) 
  }
  
  
plot<-    ggplot(var, aes(x = mean, y = bio)) +
  geom_point(colour = ifelse(var$Status=="TRUE","red","black"), size = 1.5, alpha = 1.5) + ggrepel::geom_text_repel(data = top20, mapping = aes(label = rownames(top20),x = mean,  y = bio), box.padding=unit(1, "lines"),
                                                                                                                    point.padding=unit(0.5, "lines"),
                                                                                                                    segment.colour = "purple",segment.size = 0.5,segment.alpha = 0.5,max.overlaps = Inf) +
  geom_point(data = top20, mapping = aes(label = rownames(top20)), color = "purple") + cowplot::theme_cowplot()+
  labs(x="Average Expression",y="Standardized Variance")

return(list(sce, plot))
}

```


## Workflow
# We use the TENxPBMCData function to download the relevant files from Bioconductorâ€™s ExperimentHub web resource.
```{r}
library(TENxPBMCData)
args(TENxPBMCData)
```
```{r}
tenx_pbmc4k <- TENxPBMCData(dataset = "pbmc4k")
tenx_pbmc4k
```
## QC 

```{r}
sce<- add_gene_symbol(tenx_pbmc4k)
sce<- add_QC_stats(sce)
sce
```
## QC Filter

```{r}
  
QC_Stats <- function(sce){

  min_UMI <- min(sce@colData$nCount_RNA)
  max_UMI <- max(sce@colData$nCount_RNA)
  mean_UMI <- mean(sce@colData$nCount_RNA)
  
  
  mean_gene <- mean(sce@colData$nFeature_RNA)
  min_gene <- min(sce@colData$nFeature_RNA)
  max_gene <- max(sce@colData$nFeature_RNA)
  
  min_Hemoglobin <- round(min(sce@colData$Hemoglobin_gene_percent), 2)
  max_Hemoglobin <- round(max(sce@colData$Hemoglobin_gene_percent), 2)
  mean_Hemoglobin <- round(mean(sce@colData$Hemoglobin_gene_percent), 2)
  
  min_mito <- round(min(sce@colData$Mito_gene_percent), 2)
  max_mito <- round(max(sce@colData$Mito_gene_percent), 2)
  mean_mito <- round(mean(sce@colData$Mito_gene_percent), 2)
  
  min_Ribosomal <- round(min(sce@colData$Ribosomal_gene_percent), 2)
  max_Ribosomal <- round(max(sce@colData$Ribosomal_gene_percent), 2)
  mean_Ribosomal <- round(mean(sce@colData$Ribosomal_gene_percent), 2)
  
  
  nGene_summary <- data.frame(c(min_gene, mean_gene, max_gene),
                              c(min_UMI, mean_UMI, max_UMI),
                              c(min_mito,mean_mito,max_mito),
                              c(min_Ribosomal,mean_Ribosomal, max_Ribosomal),
                              c(min_Hemoglobin, mean_Hemoglobin, max_Hemoglobin),
                              row.names = c("Min", "mean", "Max"))
  nGene_summary <- t(nGene_summary)
  row.names(nGene_summary) <- c("nGene", "nUMI","%_Mito","%_Ribosomal","%_Hemoglogin")  
  return(nGene_summary)
}
QC_Stats(sce)
```
```{r}
sce <- filter_raw_data(sce,min_genes_per_cell = 500, max_genes_per_cell = 5000, min_count_cell = 1400, max_count_cell = 4300,mt_percnt = 0.1,ribsoml_percnt = 0.1 )
```

##Preprocess data

```{r}
#Normalize and HVG
sce<- scater::logNormCounts(sce)
variable_feature <- VariableFeatures(sce, ngene = 2000, used = 'logcounts')
sce<- variable_feature[[1]]
sce
variable_feature[[2]]

```

```{r}
# Dim Reduction
sce <- scater::runPCA(sce, exprs_values=  "VGcounts", altexp= "VGcounts")
```

```{r}
scater::plotReducedDim(sce,dimred="PCA", ncomponents=3,colour_by='Individual')
```

```{r}
sce<- scater::runUMAP(sce,dimred='PCA',name ='Umap', n_dimred=20)
scater::plotReducedDim(sce,dimred="Umap",colour_by='Individual')
```

##Clustering 

```{r}
library(bluster)
colLabels(sce) <- scran::clusterCells(sce, use.dimred="Umap", BLUSPARAM=KmeansParam(centers=11))
scater::plotReducedDim(sce,dimred="Umap",colour_by='label', text_by = 'label')
```
##Cell Type Annotation

```{r}
library(SingleR)
  ref <- celldex::BlueprintEncodeData()
```




```{r}
 rst<-SingleR(test = sce, ref = ref,  clusters = sce@colData$label, labels = ref$label.fine)
```


```{r}
sce@colData$Cell_type=rst$pruned.labels[match(sce@colData$label, rownames(rst))]
```

```{r}
p<-scater::plotReducedDim(sce,dimred="Umap",colour_by='Cell_type', text_by = 'Cell_type')
p
```

```{r}
ggsave(filename = paste('tenx_pbmc4k_BlueprintEncodeData_CellType', "pdf", sep = "."), plot = p, width = 10, height = 7)
```

```{r}
saveRDS(sce, file = 'tenx_pbmc4k_cellType.rds')
```

##Annotate Cell Type using multiple datasets at a time
##Download Human reference data from celldex

```{r}
refs<-list(HumanPrimaryCellAtlas = celldex::HumanPrimaryCellAtlasData(), 
           BlueprintEncode = celldex::BlueprintEncodeData(),
           DB_ImmuneCellExpression = celldex::DatabaseImmuneCellExpressionData(),
           NovershternHematopoietic = celldex::NovershternHematopoieticData(), 
           MonacoImmune = celldex::MonacoImmuneData())
```




```{r}
#Annotate and save plot
sapply(names(refs), FUN = function(ref_idx) {
  ref<-refs[[ref_idx]]
  ref_name<-ref_idx
 
  
  # Annotation is performed on cluster-level rather than default single-cell level
  rst<-SingleR(test = sce, ref = ref,  clusters = sce@colData$label, labels = ref$label.fine)
  
  
  # SingleR assigns labels for all clusters, even for those not in the reference
  
  sce[[ref_name]]=rst$pruned.labels[match(sce@colData$label, rownames(rst))]
  
  # Download figures
  p=scater::plotReducedDim(sce,dimred="Umap",colour_by=ref_name, text_by = ref_name)
  ggsave(filename = paste(ref_name, "pdf", sep = "."), plot = p, width = 10, height = 7)
  
  #Download sce object
  saveRDS(sce, file = paste0(ref_name,'.rds'))
})
```

